### ISSUE TITLE
offline installation fails

### ISSUE THREAD

**Issue by Oldrich-Vlasic** [AzureAD/microsoft-authentication-library-for-python#198]
We are installing multiple packages including MSAL 1.3.0 on Azure-hosted Databricks cluster (runtime 6.4). Due to security, the cluster is disconnected from the internet. The packages are first collected using `pip download` and distributed to the cluster as a zip file. The installation fails with the following (omitted exact paths):

```
An error occurred while calling o16369.addIsolatedLibrary.
: org.apache.spark.SparkException: Process List(/local_disk0/path/to/bin/python, /local_disk0/path/to/bin/pip, install, --upgrade, --find-links=/local_disk0/path/to/packages, /local_disk0path/to/packages/msal-1.3.0-py2.py3-none-any.whl, --disable-pip-version-check) exited with code 1.   Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7f95dd7f9be0>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/idna/
  Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7f95dd7fd7b8>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/idna/
  Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7f95dd7fd048>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/idna/
  Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7f95dd7fd240>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/idna/
  Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.VerifiedHTTPSConnection object at 0x7f95dd7fd518>: Failed to establish a new connection: [Errno 101] Network is unreachable')': /simple/idna/
  Could not find a version that satisfies the requirement idna<2.9,>=2.5 (from requests<3,>=2.0.0->msal==1.3.0) (from versions: 2.9)
No matching distribution found for idna<2.9,>=2.5 (from requests<3,>=2.0.0->msal==1.3.0)
```

Note that the installation process succeeds on a cluster with the same configuration if it has access to internet.

Is there a way to perform the installation when fully offline?

**Comment by rayluo** [comment]
It seems this is a generic Python package installation topic not specific to MSAL Python. In particular, the error message you pasted, happened when installing `requests`, which is a popular Python package. If you can not get that installed, perhaps most of your other Python applications would not work either.

Lucky that, [this answer](https://stackoverflow.com/a/14447068) would probably help. I haven't done this myself. Let us know how it goes.

**Comment by Oldrich-Vlasic** [comment]
I'm not sure that is quite correct, what makes you think it fails while installing `requests`? The command that fails is written after `org.apache.spark.SparkException`, namely
```python pip install --upgrade --find-links=/path/to/packages /path/to/packages/msal-1.3.0-py2.py3-none-any.whl --disable-pip-version-check```

Please note that the linked answer is what we are currently doing, i.e. issuing
```python pip download azure-keyvault-secrets azure-identity```
and then providing the downloaded packages as a zip file to the cluster for installation (we do not have control over the installation process). This works perfectly fine on a cluster with access to the internet but fails without.

**Comment by rayluo** [comment]
> what makes you think it fails while installing `requests`?

The last 2 lines of your earlier error log made me think it fails while installing `requests`.

```
Could not find a version that satisfies the requirement idna<2.9,>=2.5 (from requests<3,>=2.0.0->msal==1.3.0) (from versions: 2.9)
No matching distribution found for idna<2.9,>=2.5 (from requests<3,>=2.0.0->msal==1.3.0)
```

That might surprise you, because you thought you were just installing `msal`, but it turns out `requests` is one of the dependencies needed by `msal`, and that `idna` seems like a dependency of `requests`.

Out of curiosity, were you able to install some other library (other than `msal`) but which still depends on `requests`?

> the linked answer is what we are currently doing, i.e. issuing
`python pip download azure-keyvault-secrets azure-identity`
and then providing the downloaded packages as a zip file to the cluster for installation (we do not have control over the installation process). This works perfectly fine on a cluster with access to the internet but fails without.

Like I said, I haven't personally tried that `pip download ...` and then `pip install --no-index --find-links /path/to/download/dir ...` combination. Per the wording of that [linked answer](https://stackoverflow.com/questions/11091623/how-to-install-packages-offline/14447068#14447068), it is supposed to work that way. Perhaps that is the direction to dig deeper? For example, one comment there also suggests the pip version on both machines would need to match.

This really sounds like a question for pip.

**Comment by navyasric** [comment]
Closing this as the question has been answered. Please reopen if you have further issues.

**Comment by Oldrich-Vlasic** [comment]
Sorry for late response, in the end, I do believe it was a pip issue. We looked into the dependencies of all packages and none of them specified `idna` except as `idna>=2.5` I think. Although all dependencies were downloaded, pip decided it needed `idna==2.9` for some reason and failed as it was an offline installation and this version was not present.

We have resolved the issue. Also, the original issue was not connected to MSAL. Thank you for help.
